sway-lib-std/src/experimental/storage.sw

```diff
@@ -0,0 +1,287 @@
+///
+/// Returns `Option(value)` if a the storage slots read were valid and contain `value`.
+/// Otherwise, return `None`.
+///
+/// ### Arguments
+///
+/// * `key` - The storage slot to load the value from.
+/// * `offset` - An offset, in words, from the start of slot at `key`, from which the value should
+///              be read.
+///
+/// ### Examples
+///
+/// ```sway
+/// let five = 5_u64;
+/// write(ZERO_B256, 2, five);
+/// let stored_five = read::<u64>(ZERO_B256, 2);
+/// assert(five == stored_five);
+/// ```
+#[storage(read)]
+pub fn read<T>(key: b256, offset: u64) -> Option<T> {
+    // NOTE: we are leaking this value on the heap.
+    // Get the number of storage slots needed based on the size of `T`
+    let number_of_slots = (offset * 8 + __size_of::<T>() + 31) >> 5;
+
+    // Allocate a buffer for the result. Its size needs to be a multiple of 32 bytes so we can 
+    // make the 'quad' storage instruction read without overflowing.
+    let result_ptr = alloc::<u64>(number_of_slots * 32);
+
+    // Read `number_of_slots * 32` bytes starting at storage slot `key` and return an `Option` 
+    // wrapping the value stored at `result_ptr + offset` if all the slots are valid. Otherwise, 
+    // return `Option::None`.
+    if __state_load_quad(key, result_ptr, number_of_slots) {
+        Option::Some(result_ptr.add::<u64>(offset).read::<T>())
+    } else {
+        Option::None
+    }
+}
+
+/// Clear a sequence of consecutive storage slots starting at a some key. Returns a Boolean
+/// indicating whether all of the storage slots cleared were previously set.
+///
+/// ### Arguments
+///
+/// * `key` - The key of the first storage slot that will be cleared
+///
+/// ### Examples
+///
+/// ```sway
+/// let five = 5_u64;
+/// write(ZERO_B256, 0, five);
+/// let cleared = clear::<u64>(ZERO_B256);
+/// assert(cleared);
+/// assert(read::<u64>(ZERO_B256, 0).is_none());
+/// ```
+#[storage(write)]
+pub fn clear<T>(key: b256) -> bool {
+    // Get the number of storage slots needed based on the size of `T` as the ceiling of 
+    // `__size_of::<T>() / 32`
+    let number_of_slots = (__size_of::<T>() + 31) >> 5;
+
+    // Clear `number_of_slots * 32` bytes starting at storage slot `key`.
+    __state_clear(key, number_of_slots)
+}
+
+impl<T> StorageHandle<T> {
+    /// Reads a value of type `T` starting at the location specified by `self`. If the value
+    /// crosses the boundary of a storage slot, reading continues at the following slot.
+    ///
+    /// Returns the value previously stored if a the storage slots read were
+    /// valid and contain `value`. Panics otherwise.
+    ///
+    /// ### Arguments
+    ///
+    /// None
+    ///
+    /// ### Examples
+    ///
+    /// ```sway
+    /// fn foo() {
+    ///     let r: StorageHandle<u64> = StorageHandle {
+    ///         key: 0x0000000000000000000000000000000000000000000000000000000000000000,
+    ///         offset: 2,
+    ///     };
```

Comment:
Do we want to allow users to construct arbitrary storage handles? It might limit our ability to provide some more ergonomic API for known-valid storage keys in the future? I guess that ship has already sailed though as we publicly expose the `write` and `clear` functions (without any kind of `unsafe` indicator) anyways.

Maybe we should at least keep the fields conservatively private for now and offer a `StorageHandle::from_key_and_offset` constructor, in case we decide we need to add something like a `_type: core::marker::PhantomData<T>` or similar in the future (which would otherwise be a breaking change)?

Reply:
I don't believe we have the ability to keep struct fields private at the moment, though I certainly agree with you that we shouldn't allow constructing arbitrary storage keys in the future.